modules-js/react-fleet/src/form-elements/inputs/MemorableDateInput.tsx

```diff
@@ -0,0 +1,458 @@
+  readonly earliest: Date | null;
+  readonly latest: Date | null;
+  readonly initial: Date | null;
+
+  private timeout: any;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.initial = props.initialDate ? new Date(props.initialDate) : null;
+
+    checkIfPropsValid(props);
+
+    this.earliest = setDateLimit(
+      props.earliestDate || '',
+      !!props.onlyAllowFuture
+    );
+    this.latest = setDateLimit(props.latestDate || '', !!props.onlyAllowPast);
+
+    this.state = {
+      hasFocus: false,
+      error: '',
+      date: this.initial,
+      fields: this.initial
+        ? {
+            month: this.initial.getMonth(),
+            day: this.initial.getDate(),
+            year: this.initial.getFullYear(),
+          }
+        : ({
+            month: '',
+            day: '',
+            year: '',
+          } as any),
+    };
+  }
+
+  // Determine if date range limits have been declared, and whether a given
+  // date is within that range. If not, this.state.error is set to reflect this.
+  private checkIfDateInRange(date: Date): boolean {
+    const errorString = isDateValid(
+      date,
+      this.earliest,
+      this.latest,
+      this.props.onlyAllowPast,
+      this.props.onlyAllowFuture
+    );
+
+    this.setState({ error: errorString });
+
+    return errorString.length === 0;
+  }
+
+  // If date is within range, update value. Otherwise, clear the value.
+  private updateDate(): void {
+    let date: Date | null = new Date(
+      `${this.state.fields.month}/${this.state.fields.day}/${
+        this.state.fields.year
+      }`
+    );
+
+    // If supplied date is out of range, clear the date value.
+    if (!this.checkIfDateInRange(date)) {
+      date = null;
+    }
+
+    // Update state and call parent component’s method with the new date value.
+    this.setState({ date }, () => {
+      this.props.handleDate(date);
+    });
+  }
+
+  // Focus and blur handlers are used to determine whether a user is focused
+  // on any of the fields for this component.
+  private handleFocus = (): void => {
+    clearTimeout(this.timeout);
+
+    this.setState({ hasFocus: true });
+  };
+
+  private handleBlur = (): void => {
+    this.timeout = setTimeout(() => this.setState({ hasFocus: false }), 0);
+  };
+
+  // Handler for the user-controlled input fields.
+  private handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    this.setState({
+      fields: {
+        ...this.state.fields,
+        [event.target.name]: event.target.value,
+      },
+    });
+  };
+
+  public componentDidUpdate(
+    _prevProps: Readonly<Props>,
+    prevState: Readonly<State>
+  ): void {
+    const { fields } = this.state;
+    const errorString = isInputComplete(fields);
+
+    // Only calculate a Date from the user-supplied values when a complete date
+    // has been entered by the user.
+    if (prevState.fields !== fields && errorString.length === 0) {
+      this.updateDate();
+    }
+
```

Comment:
I'm coming around a little bit on your use of `componentDidUpdate`. I think that there’s value in using it to look at `fields`, since it does mean that you can call `updateDate` in response to changes in to `fields` regardless of how they occurred (whether in `handleChange` or any other methods).

Just be mindful that you’re not writing imperative / mutation code that could be better handled as a function in `render`. For example, does `error` have to be in `State`? Or could you just calculate it in `render` from the current value of `fields` and `hasFocus`? (Or some State instead of `hasFocus`, like `touched`.)